[oneTBB] task_handle related extensions (#1)
anton-potapov/oneAPI-spec: source/elements/oneTBB/source/task_scheduler/task_group/task_handle.rst

.. SPDX-FileCopyrightText: 2019-2020 Intel Corporation
......................................................
.. SPDX-License-Identifier: CC-BY-4.0

===========
task_handle
===========
**[scheduler.task_handle]**

An instance of ``task_handle`` type owns a deferred task object.

.. code:: cpp

>    namespace tbb {

Reviewer: ```suggestion
   namespace oneapi {
   namespace tbb {
```

   
      class task_handle {
      public:
          task_handle();
          task_handle(task_handle&& src);
          
          ~task_handle();
          
          task_handle& operator=(task_handle&& src);
      
          explicit operator bool() const noexcept;
      }; 
      
      bool operator==(task_handle const& h, std::nullptr_t) noexcept;
      bool operator==(std::nullptr_t, task_handle const& h) noexcept;
      
      bool operator!=(task_handle const& h, std::nullptr_t) noexcept;
      bool operator!=(std::nullptr_t, task_handle const& h) noexcept;
                

>    } // namespace tbb

Reviewer: ```suggestion
   } // namespace tbb
   } // namespace oneapi
```

Member Functions
----------------

.. cpp:function:: task_handle()

    Creates an empty ``task_handle`` object.

.. cpp:function:: task_handle(task_handle&& src)

>     Constructs ``task_handle`` object with the content of ``src`` using move semantics. ``src`` is left in an empty state.

Reviewer: ```suggestion
    Constructs a ``task_handle`` object with the contents of ``src`` using the move semantics. ``src`` remains in an empty state.
```

Author: I believe using verb `remains` is not precise here, as ``src`` change is state from non-empty to empty after the assignment.     

what about :

 > ``src`` becomes empty after the assignment.

?

Reviewer: No objections

Author: ```suggestion
    Constructs ``task_handle`` object with the content of ``src`` using move semantics. ``src`` becomes empty after the construction. 
```

.. cpp:function:: ~task_handle()

    Destroys the ``task_handle`` object and associated task if it exists. 

.. cpp:function:: task_handle& operator=(task_handle&& src)

>     Replaces the content of ``task_handle`` object with the content of ``src`` using move semantics. ``src`` is left in an empty state.

Reviewer: ```suggestion
    Replaces the contents of the ``task_handle`` object with the contents of ``src``, using the move semantics. ``src`` remains in an empty state.
```

Author: BTW why `contents` instead of content ?

Author: ```suggestion
    Replaces the content of ``task_handle`` object with the content of ``src`` using move semantics. ``src`` becomes empty after the assignment.
```

>     The previously associated task object, if any, is destroyed before the assignment. 

Reviewer: ```suggestion
    The previously associated task object, if any, is destroyed before being assigned.  
```

Author: an instance of ``task_handle`` is being assigned, not the task object (which is associated with the ``task_handle``)

Author: please mark the conversation as resolved, if there are no objections

Reviewer: ok

    **Returns:** Reference to ``*this``.

.. cpp:function:: explicit operator bool() const noexcept

    Checks if ``*this`` has an associated task object.

    **Returns:** ``true`` if ``*this`` is not empty, ``false`` otherwise.

Non-Member Functions
--------------------

.. code:: cpp

    bool operator==(task_handle const& h, std::nullptr_t) noexcept
    bool operator==(std::nullptr_t, task_handle const& h) noexcept

>     **Returns**: ``true`` if ``h`` is empty, ``false`` otherwise.

Reviewer: ```suggestion
**Returns**: ``true`` if ``h`` is empty, ``false`` otherwise.
```

Reviewer: rst syntax

Author: how does this differ from the snippet above (or the entries in the task_group_cls.rst )?

.. code:: cpp

    bool operator!=(task_handle const& h, std::nullptr_t) noexcept
    bool operator!=(std::nullptr_t, task_handle const& h) noexcept

>     **Returns**: ``true`` if ``h`` is not empty, ``false`` otherwise.

Reviewer: ```suggestion
**Returns**: ``true`` if ``h`` is not empty, ``false`` otherwise.
```
